### PR TITLE
2565 gift email and postcode validation

### PR DESCRIPTION
**Pull request by alexflav23**
## Why are you doing this?

Fixing a SalesForce bug where step functions are failing as part of the weekly checkout flow because we do not validate gift recipient email addresses. This causes an infinite loop on the step function which ends up re-submitting the checkout request every time it fails. We now validate against malformed emails at submission time, while still allowing for gift recipients where no email address is specified.

[**Trello Card**](https://trello.com/c/ONmjZVen/2565-clientside-validate-checkout-fields-to-avoid-salesforce-errors)

## Changes

* Added a function to validate optional email addresses
* Added a reducer and a validation rule against the gift recipient email address on the weekly checkout flow.
* Fix Scala warning against using deprecated implicit collection conversions, favouring new API based on implicit value classes with explicit calls.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -77,6 +77,7 @@ lazy val `support-frontend` = (project in file("support-frontend"))
     buildInfoPackage := "app",
     buildInfoOptions += BuildInfoOption.ToMap,
     scalastyleFailOnError := true,
+    fork in run := true,
```
Comment (johnduffell) [inline build.sbt]: If it's likely to have unexpected side effects (probably not) or clutter the main pr too much, it can be worth putting in a separate PR.
If its in this PR it would be good to write a brief comment here (in github is fine) explaining why it's useful/justification and any potential downsides.  That helps reviewers know that it was intentional

File: support-frontend/assets/helpers/formValidation.js [hunk 2]
```diff
@@ -13,6 +13,10 @@ import { Canada, UnitedStates } from './internationalisation/countryGroup';
 export const emailRegexPattern = '^[a-zA-Z0-9\\.!#$%&\'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$';
 
 export const isNotEmpty: (string | null) => boolean = input => !!input && input.trim() !== '';
+
+export const isEmpty: (string | null) => boolean = input =>
+  typeof input === 'undefined' || input == null || input.trim().length === 0;
```
Comment (johnduffell) [inline support-frontend/assets/helpers/formValidation.js]: javascript 😂

File: support-frontend/assets/helpers/subscriptionsForms/validation.js [hunk 3]
```diff
@@ -27,7 +27,9 @@ function notNull<A>(value: A): boolean {
 
 function firstError<FieldType>(field: FieldType, errors: FormError<FieldType>[]): Option<string> {
 
+
   const msgs = errors.filter(err => err.field === field).map(err => err.message);
+
```
Comment (johnduffell) [inline support-frontend/assets/helpers/subscriptionsForms/validation.js]: in general better to avoid relayout in parts of the code that weren't otherwise changed

File: support-frontend/assets/helpers/formValidation.js [hunk 4]
```diff
@@ -13,6 +13,10 @@ import { Canada, UnitedStates } from './internationalisation/countryGroup';
 export const emailRegexPattern = '^[a-zA-Z0-9\\.!#$%&\'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$';
 
 export const isNotEmpty: (string | null) => boolean = input => !!input && input.trim() !== '';
+
+export const isEmpty: (string | null) => boolean = input =>
```
Comment (rupertbates) [inline support-frontend/assets/helpers/formValidation.js]: We should be able to redefine `isNotEmpty` as `! isEmpty()` now shouldn't we?

Comment (alexflav23) [reply support-frontend/assets/helpers/formValidation.js]: @rupertbates That was my first point of call, but we figured make it explicit.

Comment (johnduffell) [reply support-frontend/assets/helpers/formValidation.js]: actually rupert is suggesting something different, we originally had !isNotEmpty and isNotEmpty, he is suggesting isEmpty and !isEmpty

File: support-frontend/assets/helpers/subscriptionsForms/rules.js [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import { formError, nonEmptyString, notNull, validate } from './validation';
 import type { FormField, FormFields } from './formFields';
 import type { FormError } from './validation';
+import {checkOptionalEmail} from "../formValidation";
```
[the remark this answers is not in the document]
Comment (rupertbates) [reply support-frontend/assets/helpers/subscriptionsForms/rules.js]: I think it was just right clicking on .eslintrc and choosing 'Apply ESLint code rules' wasn't it?

Screenshot 2019-10-08 at 17 14 36

**Comment by mattoh1** [comment]
Great first PR @alexflav23 👍

File: support-frontend/assets/helpers/subscriptionsForms/validation.js [hunk 6]
```diff
@@ -26,10 +26,8 @@ function notNull<A>(value: A): boolean {
 // ----- Functions ----- //
 
 function firstError<FieldType>(field: FieldType, errors: FormError<FieldType>[]): Option<string> {
-
   const msgs = errors.filter(err => err.field === field).map(err => err.message);
   return headOption(msgs);
-
```
Comment (johnduffell) [inline support-frontend/assets/helpers/subscriptionsForms/validation.js]: don't worry this time, but generally, watch out for random reformatting noise
